File: controllers/report_excel.mjs
import fs from "node:fs/promises";
import { SpreadsheetFile, Workbook } from "@oai/artifact-tool";

const [inputPath, outputPath] = process.argv.slice(2);
if (!inputPath || !outputPath) throw new Error("Se requiere el archivo de datos y la ruta de salida.");
const preview = JSON.parse(await fs.readFile(inputPath, "utf8"));
const workbook = Workbook.create();
const sheet = workbook.worksheets.add("Informe");
const rows = preview.datos || [];
const columns = rows.length ? Object.keys(rows[0]) : ["Resultado"];
const lastColumn = String.fromCharCode(64 + Math.min(columns.length, 26));

sheet.showGridLines = false;
sheet.mergeCells(`A1:${lastColumn}1`);
sheet.getRange("A1").values = [["BUSCA HUELLAS"]];
sheet.getRange("A1").format = { fill: "#0F5238", font: { bold: true, color: "#FFFFFF", size: 16 }, horizontalAlignment: "center", verticalAlignment: "center" };
sheet.getRange("A1").format.rowHeight = 30;
sheet.mergeCells(`A2:${lastColumn}2`);
sheet.getRange("A2").values = [[preview.titulo]];
sheet.getRange("A2").format = { font: { bold: true, color: "#0F5238", size: 14 }, horizontalAlignment: "center" };
sheet.mergeCells(`A3:${lastColumn}3`);
const dates = preview.fecha_inicio || preview.fecha_fin ? `Periodo: ${preview.fecha_inicio || "Inicio"} a ${preview.fecha_fin || "Hoy"}` : "Sin filtro de fechas";
sheet.getRange("A3").values = [[`${preview.tipo.replaceAll("_", " ")} - ${dates}`]];
sheet.getRange("A3").format = { font: { color: "#56615A", italic: true }, horizontalAlignment: "center" };

sheet.getRange(`A5:${lastColumn}5`).values = [columns.map((column) => column.replaceAll("_", " ").replace(/\b\w/g, (letter) => letter.toUpperCase()))];
sheet.getRange(`A5:${lastColumn}5`).format = { fill: "#0F5238", font: { bold: true, color: "#FFFFFF" }, horizontalAlignment: "center", wrapText: true };
if (rows.length) {
  sheet.getRangeByIndexes(5, 0, rows.length, columns.length).values = rows.map((row) => columns.map((column) => row[column] ?? ""));
  const dataRange = sheet.getRangeByIndexes(4, 0, rows.length + 1, columns.length);
  dataRange.format.borders = { preset: "all", style: "thin", color: "#D8DED7" };
  sheet.tables.add(`A5:${lastColumn}${rows.length + 5}`, true, "InformeDatos");
} else {
  sheet.mergeCells(`A6:${lastColumn}6`);
  sheet.getRange("A6").values = [["No se encontraron resultados con los filtros seleccionados."]];
  sheet.getRange("A6").format = { horizontalAlignment: "center", font: { color: "#56615A", italic: true } };
}
sheet.getRange(`A1:${lastColumn}${Math.max(6, rows.length + 5)}`).format.autofitColumns();
for (let index = 0; index < columns.length; index += 1) {
  const column = sheet.getRangeByIndexes(0, index, Math.max(6, rows.length + 5), 1);
  if (column.format.columnWidth > 28) column.format.columnWidth = 28;
  if (column.format.columnWidth < 13) column.format.columnWidth = 13;
}
sheet.freezePanes.freezeRows(5);

const output = await SpreadsheetFile.exportXlsx(workbook);
await output.save(outputPath);

File: controllers/report_excel.py

"""Generación de libros Excel para los informes administrativos.

No depende de Node ni de paquetes instalados fuera del proyecto.  Esto es
importante porque los informes se generan en el servidor de la aplicación,
no en el entorno de desarrollo de Codex.
"""

import os
import uuid
import zipfile
from xml.sax.saxutils import escape

from controllers.report_pdf import REPORTS_FOLDER


def crear_excel_informe(preview):
    """Crea un .xlsx válido y devuelve el nombre seguro para almacenar en BD."""
    os.makedirs(REPORTS_FOLDER, exist_ok=True)
    nombre_archivo = f"informe_{uuid.uuid4().hex}.xlsx"
    ruta_salida = os.path.join(REPORTS_FOLDER, nombre_archivo)
    filas = preview.get("datos") or []
    columnas = list(filas[0].keys()) if filas else ["Resultado"]
    periodo = (
        f"Periodo: {preview.get('fecha_inicio') or 'Inicio'} a {preview.get('fecha_fin') or 'Hoy'}"
        if preview.get("fecha_inicio") or preview.get("fecha_fin")
        else "Sin filtro de fechas"
    )
    contenido = [
        ["BUSCA HUELLAS"],
        [preview.get("titulo") or "Informe personalizado"],
        [f"{str(preview.get('tipo') or '').replace('_', ' ')} - {periodo}"],
        [],
        [str(columna).replace("_", " ").title() for columna in columnas],
    ]
    if filas:
        contenido.extend([[fila.get(columna, "") for columna in columnas] for fila in filas])
    else:
        contenido.append(["No se encontraron resultados con los filtros seleccionados."])

    try:
        with zipfile.ZipFile(ruta_salida, "w", zipfile.ZIP_DEFLATED) as libro:
            libro.writestr("[Content_Types].xml", _CONTENT_TYPES)
            libro.writestr("_rels/.rels", _RELS)
            libro.writestr("xl/workbook.xml", _WORKBOOK)
            libro.writestr("xl/_rels/workbook.xml.rels", _WORKBOOK_RELS)
            libro.writestr("xl/styles.xml", _STYLES)
            libro.writestr("xl/worksheets/sheet1.xml", _hoja_xml(contenido, len(columnas), bool(filas)))
    except Exception:
        if os.path.isfile(ruta_salida):
            os.remove(ruta_salida)
        raise
    return nombre_archivo


def _celda(valor, fila, columna, estilo=0):
    referencia = f"{_columna_excel(columna)}{fila}"
    texto = "" if valor is None else str(valor)
    return f'<c r="{referencia}" s="{estilo}" t="inlineStr"><is><t>{escape(texto)}</t></is></c>'


def _columna_excel(numero):
    resultado = ""
    while numero:
        numero, resto = divmod(numero - 1, 26)
        resultado = chr(65 + resto) + resultado
    return resultado


def _hoja_xml(contenido, total_columnas, tiene_datos):
    ultima_columna = _columna_excel(total_columnas)
    filas_xml = []
    for numero_fila, valores in enumerate(contenido, start=1):
        if numero_fila in (1, 2, 3):
            filas_xml.append(f'<row r="{numero_fila}"><c r="A{numero_fila}" s="{numero_fila - 1}" t="inlineStr"><is><t>{escape(str(valores[0]))}</t></is></c></row>')
        elif numero_fila == 5:
            filas_xml.append(f'<row r="5">{"".join(_celda(valor, numero_fila, indice, 3) for indice, valor in enumerate(valores, 1))}</row>')
        else:
            filas_xml.append(f'<row r="{numero_fila}">{"".join(_celda(valor, numero_fila, indice, 4 if tiene_datos else 5) for indice, valor in enumerate(valores, 1))}</row>')
    columnas = "".join(f'<col min="{indice}" max="{indice}" width="22" customWidth="1"/>' for indice in range(1, total_columnas + 1))
    merges = f'<mergeCells count="3"><mergeCell ref="A1:{ultima_columna}1"/><mergeCell ref="A2:{ultima_columna}2"/><mergeCell ref="A3:{ultima_columna}3"/></mergeCells>'
    return f'''<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<worksheet xmlns="http://schemas.openxmlformats.org/spreadsheetml/2006/main"><sheetViews><sheetView workbookViewId="0"><pane ySplit="5" topLeftCell="A6" activePane="bottomLeft" state="frozen"/></sheetView></sheetViews><cols>{columnas}</cols><sheetData>{''.join(filas_xml)}</sheetData>{merges}</worksheet>'''


_CONTENT_TYPES = '''<?xml version="1.0" encoding="UTF-8" standalone="yes"?><Types xmlns="http://schemas.openxmlformats.org/package/2006/content-types"><Default Extension="rels" ContentType="application/vnd.openxmlformats-package.relationships+xml"/><Default Extension="xml" ContentType="application/xml"/><Override PartName="/xl/workbook.xml" ContentType="application/vnd.openxmlformats-officedocument.spreadsheetml.sheet.main+xml"/><Override PartName="/xl/worksheets/sheet1.xml" ContentType="application/vnd.openxmlformats-officedocument.spreadsheetml.worksheet+xml"/><Override PartName="/xl/styles.xml" ContentType="application/vnd.openxmlformats-officedocument.spreadsheetml.styles+xml"/></Types>'''
_RELS = '''<?xml version="1.0" encoding="UTF-8" standalone="yes"?><Relationships xmlns="http://schemas.openxmlformats.org/package/2006/relationships"><Relationship Id="rId1" Type="http://schemas.openxmlformats.org/officeDocument/2006/relationships/officeDocument" Target="xl/workbook.xml"/></Relationships>'''
_WORKBOOK = '''<?xml version="1.0" encoding="UTF-8" standalone="yes"?><workbook xmlns="http://schemas.openxmlformats.org/spreadsheetml/2006/main" xmlns:r="http://schemas.openxmlformats.org/officeDocument/2006/relationships"><sheets><sheet name="Informe" sheetId="1" r:id="rId1"/></sheets></workbook>'''
_WORKBOOK_RELS = '''<?xml version="1.0" encoding="UTF-8" standalone="yes"?><Relationships xmlns="http://schemas.openxmlformats.org/package/2006/relationships"><Relationship Id="rId1" Type="http://schemas.openxmlformats.org/officeDocument/2006/relationships/worksheet" Target="worksheets/sheet1.xml"/><Relationship Id="rId2" Type="http://schemas.openxmlformats.org/officeDocument/2006/relationships/styles" Target="styles.xml"/></Relationships>'''
_STYLES = '''<?xml version="1.0" encoding="UTF-8" standalone="yes"?><styleSheet xmlns="http://schemas.openxmlformats.org/spreadsheetml/2006/main"><fonts count="5"><font><sz val="11"/><name val="Calibri"/></font><font><b/><color rgb="FFFFFFFF"/><sz val="16"/><name val="Calibri"/></font><font><b/><color rgb="FF0F5238"/><sz val="14"/><name val="Calibri"/></font><font><b/><color rgb="FFFFFFFF"/><name val="Calibri"/></font><font><name val="Calibri"/></font></fonts><fills count="3"><fill><patternFill patternType="none"/></fill><fill><patternFill patternType="gray125"/></fill><fill><patternFill patternType="solid"><fgColor rgb="FF0F5238"/><bgColor indexed="64"/></patternFill></fill></fills><borders count="2"><border/><border><left style="thin"><color rgb="FFD8DED7"/></left><right style="thin"><color rgb="FFD8DED7"/></right><top style="thin"><color rgb="FFD8DED7"/></top><bottom style="thin"><color rgb="FFD8DED7"/></bottom></border></borders><cellStyleXfs count="1"><xf numFmtId="0" fontId="0" fillId="0" borderId="0"/></cellStyleXfs><cellXfs count="6"><xf numFmtId="0" fontId="0" fillId="0" borderId="0" xfId="0"/><xf numFmtId="0" fontId="1" fillId="2" borderId="0" applyFont="1" applyFill="1" applyAlignment="1"><alignment horizontal="center" vertical="center"/></xf><xf numFmtId="0" fontId="2" fillId="0" borderId="0" applyFont="1" applyAlignment="1"><alignment horizontal="center"/></xf><xf numFmtId="0" fontId="3" fillId="2" borderId="1" applyFont="1" applyFill="1" applyBorder="1" applyAlignment="1"><alignment horizontal="center" wrapText="1"/></xf><xf numFmtId="0" fontId="4" fillId="0" borderId="1" applyBorder="1" applyAlignment="1"><alignment vertical="top" wrapText="1"/></xf><xf numFmtId="0" fontId="4" fillId="0" borderId="0" applyAlignment="1"><alignment horizontal="center"/></xf></cellXfs></styleSheet>'''

"""Generación de libros Excel para los informes administrativos."""

import json
import os
import subprocess
import sys
import uuid

from controllers.report_pdf import REPORTS_FOLDER, ruta_pdf_informe


def crear_excel_informe(preview):
    os.makedirs(REPORTS_FOLDER, exist_ok=True)
    nombre_archivo = f"informe_{uuid.uuid4().hex}.xlsx"
    ruta_salida = os.path.join(REPORTS_FOLDER, nombre_archivo)
    ruta_entrada = os.path.join(REPORTS_FOLDER, f".informe_{uuid.uuid4().hex}.json")
    try:
        with open(ruta_entrada, "w", encoding="utf-8") as archivo:
            json.dump(preview, archivo, ensure_ascii=False)
        resultado = subprocess.run(
            ["node", os.path.join(os.path.dirname(__file__), "report_excel.mjs"), ruta_entrada, ruta_salida],
            capture_output=True,
            text=True,
            timeout=45,
            check=False,
        )
        if resultado.returncode != 0 or not os.path.isfile(ruta_salida):
            raise RuntimeError(resultado.stderr.strip() or "No se pudo crear el archivo Excel.")
        return nombre_archivo
    finally:
        if os.path.isfile(ruta_entrada):
            os.remove(ruta_entrada)
